plugins/synaprom/fu-synaprom-device.c

```diff
@@ -0,0 +1,504 @@
+	buf = g_bytes_get_data (fw, &sz);
+	while (sz != 0) {
+		guint32 chunksz;
+		g_autoptr(GByteArray) chunk = g_byte_array_new ();
+
+		/* get chunk size */
+		if (sz < sizeof(guint32)) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_NOT_SUPPORTED,
+					     "No enough data for patch len");
+			return FALSE;
+		}
+		memcpy (&chunksz, buf, sizeof(guint32));
+		buf += sizeof(guint32);
+		sz -= sizeof(guint32);
+		if (sz < chunksz) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_NOT_SUPPORTED,
+					     "No enough data for patch chunk");
+			return FALSE;
+		}
+
+		/* download chunk */
+		g_byte_array_append (chunk, buf, chunksz);
+		if (!fu_synaprom_device_cmd_download_chunk (self, chunk, error))
+			return FALSE;
+
+		/* next chunk */
+		buf += chunksz;
+		sz -= chunksz;
+	}
+
+	/* success! */
+	fu_device_set_progress (device, 100);
+	return TRUE;
+}
+
+static gboolean
+fu_synaprom_device_probe (FuUsbDevice *device, GError **error)
+{
+	fu_device_add_flag (FU_DEVICE (device), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_remove_delay (FU_DEVICE (device), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	fu_device_set_name (FU_DEVICE (device), "Prometheus");
+	fu_device_set_summary (FU_DEVICE (device), "Fingerprint reader");
+	fu_device_set_vendor (FU_DEVICE (device), "Synaptics");
+	fu_device_add_icon (FU_DEVICE (device), "touchpad-disabled");
+	return TRUE;
+}
+
+static gboolean
+fu_synaprom_device_attach (FuDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	gboolean ret;
+	gsize actual_len = 0;
+	guint8 data[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+
+	ret = g_usb_device_control_transfer (usb_device,
+					     G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					     G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					     G_USB_DEVICE_RECIPIENT_DEVICE,
+					     FU_SYNAPROM_USB_CTRLREQUEST_VENDOR_WRITEDFT,
+					     0x0000,
+					     0x0000,
+					     data,
+					     sizeof(data),
+					     &actual_len,
+					     2000,
+					     NULL,
+					     error);
+	if (!ret)
+		return FALSE;
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	if (!g_usb_device_reset (usb_device, error)) {
+		g_prefix_error (error, "failed to force-reset device: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_synaprom_device_detach (FuDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	gboolean ret;
+	gsize actual_len = 0;
+	guint8 data[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00 };
+
+	ret = g_usb_device_control_transfer (usb_device,
+					     G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					     G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					     G_USB_DEVICE_RECIPIENT_DEVICE,
+					     FU_SYNAPROM_USB_CTRLREQUEST_VENDOR_WRITEDFT,
+					     0x0000,
+					     0x0000,
+					     data,
+					     sizeof(data),
+					     &actual_len,
```

Comment:
Shouldn't you also validate `actual_len` here?

Reply:
Fixed, thanks.